index.js

```diff
@@ -1,29 +1,67 @@
-function classNames() {
+'use strict';
+
+var SPACE = /\s+/;
+var toString = Object.prototype.toString;
+
+function classNames () {
 	var classes = '';
-	var arg;
+	var i = 0;
+	var arg = arguments[i];
+	var k;
+	var v;
 
-	for (var i = 0; i < arguments.length; i++) {
-		arg = arguments[i];
+	for (; i < arguments.length; arg = arguments[++i]) {
 		if (!arg) {
 			continue;
 		}
 
 		if ('string' === typeof arg || 'number' === typeof arg) {
 			classes += ' ' + arg;
-		} else if (Object.prototype.toString.call(arg) === '[object Array]') {
+		} else if (toString.call(arg) === '[object Array]') {
 			classes += ' ' + classNames.apply(null, arg);
 		} else if ('object' === typeof arg) {
-			for (var key in arg) {
-				if (!arg.hasOwnProperty(key) || !arg[key]) {
+			for (k in arg) {
+				if (!arg.hasOwnProperty(k)) {
 					continue;
 				}
-				classes += ' ' + key;
+				v = arg[k];
+				if (!v) {
+					classes = classes.replace(new RegExp(' ' + k, 'g'), '');
```

Comment:
You need a better regexp. This is going to fail this test:
```js
assert.equal(classnames("foo", "foobar", "bar", {foo: false}), "foobar")); 
# yields "ar bar" instead of "foobar bar"
```